File: backend-app/src/main/java/com/backend_app/backend_app/service/CategoriaService.java
package com.backend_app.backend_app.service;

import com.backend_app.backend_app.domain.Categoria;
import com.backend_app.backend_app.dto.CategoriaDTO;
import com.backend_app.backend_app.dto.UsuarioDTO;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import java.util.List;
import java.util.Optional;

public interface CategoriaService {

    List<CategoriaDTO> listadoCategorias();

    Optional<CategoriaDTO> buscarCategoriaPorId(long idCategoria);

    CategoriaDTO convertirADTO(Categoria categoria);

    void registrarCategoria(CategoriaDTO categoriaDTO);

    void editarCategoria(CategoriaDTO categoriaDTO);

    void eliminarCategoria(CategoriaDTO categoriaDTO);

    Page<CategoriaDTO> findAll(Pageable pageable);

    List<CategoriaDTO> buscarCategoriaPorNombre(String nombreCategoria);

    List<CategoriaDTO> listadoCategoriasCompletas();

}

File: frontend-app/src/app/services/categoria.service.ts
import { HttpClient, HttpParams } from "@angular/common/http";
import { Injectable } from "@angular/core";
import { Observable } from "rxjs";
import { Categoria } from "../domain/Categoria";
import { Usuario } from "../domain/Usuario";


@Injectable({
    providedIn: 'root'
})
export class CategoriaService{

    private url:string = 'http://localhost:8080/api/categorias';

    constructor(private http: HttpClient){

    }

    listadoCagtegorias(): Observable<Categoria[]>{
        return this.http.get<Categoria[]>(this.url);
    }
    
        
    buscarCategoriaPorId(id: number): Observable<Categoria>{
        return this.http.get<Categoria>(`${this.url}/${id}`);
    }
    
    guardarCategoria(categoria: Categoria){
        return this.http.post('http://localhost:8080/api/categorias/guardar', categoria);
    }

    editarCategoria(categoria: Categoria) {
        return this.http.put(`${this.url}/actualizar`, categoria);
    }

    eliminarCategoria(idCategoria: number){
        return this.http.delete(`${this.url}/${idCategoria}`);
    }

    listadoCategoriaPaginacion(page: number): Observable<any>{
        return this.http.get<any[]>(`${this.url}/page/${page}`);
    }

    
    buscarCategoriaPorNombre(nombre: string): Observable<Categoria[]>{
        const params = new HttpParams().set('nombreCategoria', nombre);
                    
        return this.http.get<Categoria[]>(`${this.url}/buscar`, { params });
    }

    listadoCategoriasCompletas(): Observable<Categoria[]>{
        return this.http.get<Categoria[]>(`${this.url}/completas`);
    }

}